### PR TITLE
Add function resolution to SPI

### LINKED ISSUE #8: Function and type namespaces
# Goals

* Allow for coexistence of functions and types with similar names but different semantics (e.g., compatibility layer for other databases like Teradata, Oracle, MySQL, Vertica, PostgreSQL, etc)
* Allow for tighter control of what functions are available in a given deployment
* Make it possible to implement user-defined functions and types in the future

# Features

* Support for `CURRENT_PATH`, `SET PATH`
* Configurable/mountable function and type plugins

# Considerations and open questions

* Views need to record CURRENT_PATH in definition
* Need to define migration strategy for re-packaging existing function library
* Function should be resolved to a handle during analysis just like tables.
* Function resolution should be transaction aware.
* Function metadata can be fetch using a handle on the coordinator.
* Function implementation can be retrieved using the handle on coordinator and workers.
* Optimizers can lookup functions using fully a specified signature.
* Function handles should be transported to worker without requiring a full rewrite of the plan IR.
* SPI interface should be easy to implement.

# TODOs

- [x] Introduce proper function abstractions to engine (FunctionManager, FunctionHandle, etc). #196
- [x] Simplify function resolution
  * Extract resolution logic from FunctionRegistry.
  * Resolve functions using simpler metadata interface.
  * Move resolution to be inside analyzer (like table resolution)
- [x] Simplify function instance usage
  * Move call convention resolution to be inside of FunctionManager
  * Callers should really only need to deal with conventions and method handles
- [x] Finish SQL PATH implementation
  * Move existing function registration to system catalog
  * Add namespace abstraction
  * Update function resolution to understand path semantics
- [ ] Update connector SPI to support function resolution
- [ ] Improve function handle transport to workers

# References

> ### 4.34 SQL-paths
> An SQL-path is a list of one or more `<schema name>`s that determines the search order for one of the following:
> — The subject routine of a `<routine invocation>` whose `<routine name>` does not contain a <schema name>.
> — The user-defined type when the `<path-resolved user-defined type name>` does not contain a `<schema name>`.

> ### 4.43.3 SQL-session properties
> An SQL-session has an SQL-path that is used to effectively qualify unqualified `<routine name>`s [...]
> The SQL-path is initially set to an implementation-de ned value, but can subsequently be changed by the successful execution of a `<set path statement>`.
> The text defining the SQL-path can be referenced by using the `<general value specification> CURRENT_PATH`.

> ### 6.4 `<value specification>` and `<target specification>`
> 4) The declared type of `CURRENT_USER`, `CURRENT_ROLE`, `SESSION_USER`, `SYSTEM_USER`, `CURRENT_CATALOG`, `CURRENT_SCHEMA`, and `CURRENT_PATH` is character string. Whether the character string is fixed-length or variable-length, and its length if it is fixed-length or maximum length if it is variable-length, are implementation-de ned. The character set of the character string is `SQL_IDENTIFIER`. The declared type collation is the character set collation of `SQL_IDENTIFIER`, and the collation derivation is implicit.
> 12) The value specified by `CURRENT_PATH` is a `<schema name list>` where `<catalog name>`s are `<delimited identifier>`s and the `<unqualified schema name>`s are `<delimited identifier>`s. Each` <schema name>` is separated from the preceding `<schema name>` by a `<comma>` with no intervening `<space>`s. The schemas referenced in this `<schema name list>` are those referenced in the SQL-path of the current SQL-session context, in the order in which they appear in that SQL-path.

> ### 10.3 `<path specification>`
> #### Function
> Specify an order for searching for an SQL-invoked routine.
> #### Format
> ```
> <path specification> ::= PATH <schema name list>
> <schema name list> ::= <schema name> [ { <comma> <schema name> }... ]
> <schema name> ::= [ <catalog name> <period> ] <unqualified schema name>
> <unqualified schema name> ::= <identifier>
> <catalog name> ::= <identifier>
> ```
> #### Syntax Rules
> 1) No two `<schema name>`s contained in `<schema name list>` shall be equivalent.

> ### 19.8 `<set path statement>`
> #### Function
> Set the SQL-path used to determine the subject routine of `<routine invocation>`s with unqualified `<routine name>`s in `<preparable statement>`s that are prepared in the current SQL-session by an `<execute immediate statement>` or a `<prepare statement>` and in `<direct SQL statement>`s that are invoked directly. The SQL-path remains the current SQL-path of the SQL-session until another SQL-path is successfully set.
> #### Format
> ```
> <set path statement> ::= SET <SQL-path characteristic>
> <SQL-path characteristic> ::= PATH <value specification>
> ```
> #### Syntax Rules
> 1) The declared type of the <value specification> shall be a character string type.
> #### Access Rules
> None.
> #### General Rules
> 1) Let `S` be `<value specification>` and let `V` be the character string that is the value of `TRIM ( BOTH ' ' FROM S )`
> a) If `V` does not conform to the Format and Syntax Rules of a `<schema name list>`, then an exception condition is raised: *invalid schema name list specification*.
> b) The SQL-path of the current SQL-session is set to `V`.
> NOTE 724 — A `<set path statement>` that is executed between a `<prepare statement>` and an `<execute statement>` has no effect on the prepared statement.
> #### Conformance Rules
> 1) Without Feature S071, “SQL paths in function and type name resolution”, Conforming SQL language shall not contain a `<set path statement>`.

### PR DESCRIPTION
**Pull request by dain**
## Description

This is the last step in adding dynamic function resolution to the SPI.  The function system has been updated in previous PRs, and this PR is focused on:
* Cleanup function APIs: rename classes, add builders, and such
* Move function APIs to the SPI package
* Wire function resolution into connector metadata

## Related issues, pull requests, and links

* Fixes #8

## Documentation

(X) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(X) Release notes entries required with the following suggested text:

```markdown
# SPI
* Add function resolution to connector. ({issue}`8`)
```

### CODE REVIEW

**Comment by dain** [comment]
@findepi yes

**Comment by wmoustafa** [comment]
This is awesome to see. Thank you so much @dain for working on this. We will check it out to ensure backward compatibility as we leveraged the internal implementation (pre-SPI) in Transport ([Ref](https://github.com/linkedin/transport/blob/master/transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java)). @phd3 FYI.

File: core/trino-main/src/main/java/io/trino/metadata/QualifiedFunctionName.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    public static QualifiedFunctionName of(String catalogName, String schemaName, String functionName)
+    {
+        return new QualifiedFunctionName(Optional.of(catalogName), Optional.of(schemaName), functionName);
+    }
+
+    private QualifiedFunctionName(Optional<String> catalogName, Optional<String> schemaName, String functionName)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/QualifiedFunctionName.java]: `requireNonNull` not needed since `catalogName` is dereferenced below (the JVM will produce a better error message in that case)

File: core/trino-main/src/main/java/io/trino/metadata/QualifiedFunctionName.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        catalogName.ifPresent(name -> checkArgument(!name.isEmpty(), "catalogName is empty"));
+        this.schemaName = requireNonNull(schemaName, "schemaName is null");
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/QualifiedFunctionName.java]: Same here

File: core/trino-main/src/main/java/io/trino/metadata/FunctionBundle.java [hunk 3]
```diff
@@ -27,7 +27,7 @@
 
     FunctionDependencyDeclaration getFunctionDependencies(FunctionId functionId, BoundSignature boundSignature);
 
-    FunctionInvoker getScalarFunctionInvoker(
+    ScalarImplementation getScalarFunctionInvoker(
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/FunctionBundle.java:37]: getScalarFunctionImplementation

File: core/trino-main/src/main/java/io/trino/metadata/ScalarImplementation.java [hunk 4]
```diff
@@ -21,13 +21,13 @@
 
 import static java.util.Objects.requireNonNull;
 
-public class FunctionInvoker
+public class ScalarImplementation
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/ScalarImplementation.java]: Maybe ScalarFunctionImplementation?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/ParametricAggregation.java [hunk 5]
```diff
@@ -146,40 +145,49 @@ public AggregationImplementation specialize(BoundSignature boundSignature, Funct
...
+        builder.inputFunction(normalizeInputMethod(
+                bindDependencies(
+                        concreteImplementation.getInputFunction(),
+                        concreteImplementation.getInputDependencies(),
+                        functionBinding,
+                        functionDependencies),
+                boundSignature,
+                inputParameterKinds));
+        concreteImplementation.getRemoveInputFunction()
+                .map(removeInputFunction -> bindDependencies(removeInputFunction,
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/ParametricAggregation.java:169]: Argument on next line

File: core/trino-main/src/main/java/io/trino/metadata/FunctionInvoker.java [hunk 6]
```diff
@@ -55,4 +48,41 @@ public List<Class<?>> getLambdaInterfaces()
...
+
+    public static class Builder
+    {
+        private MethodHandle methodHandle;
+        private MethodHandle instanceFactory;
+        private List<Class<?>> lambdaInterfaces = ImmutableList.of();
+
+        private Builder() {}
+
+        public Builder setMethodHandle(MethodHandle methodHandle)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/FunctionInvoker.java]: `withMethodHandle(...)` or just `methodHandle(...)`? Same for the ones below. BTW, in the commit to add a builder for AggregationImplementation you use the latter convention.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 7]
```diff
@@ -2025,7 +2029,17 @@ public List<GrantInfo> listTablePrivileges(Session session, QualifiedTablePrefix
     @Override
     public Collection<FunctionMetadata> listFunctions(Session session)
     {
-        return functions.listFunctions();
+        ImmutableList.Builder<FunctionMetadata> functions = ImmutableList.builder();
+        functions.addAll(this.functions.listFunctions());
+        for (SqlPathElement sqlPathElement : session.getPath().getParsedPath()) {
+            String catalog = sqlPathElement.getCatalog().map(Identifier::getValue).or(session::getCatalog)
+                    .orElseThrow(() -> new IllegalArgumentException("Session default catalog must be set to resolve a partial function name: " + null));
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:2036]: Throw a TrinoException. Also, the `+ null` at the end of the message seems like a mistake.

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 8]
```diff
@@ -2124,27 +2143,45 @@ public ResolvedFunction getCoercion(Session session, OperatorType operatorType,
...
+        return resolve(session, catalogFunctionBinding);
     }
 
-    private ResolvedFunction resolve(Session session, FunctionBinding functionBinding)
+    private ResolvedFunction resolve(Session session, CatalogFunctionBinding functionBinding)
     {
-        FunctionDependencyDeclaration declaration = functions.getFunctionDependencies(functionBinding.getFunctionId(), functionBinding.getBoundSignature());
-        FunctionMetadata functionMetadata = getFunctionMetadata(functionBinding.getFunctionId(), functionBinding.getBoundSignature());
-        return resolve(session, functionBinding, functionMetadata, declaration);
+        FunctionDependencyDeclaration declaration = getFunctionDependencyDeclaration(
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:2160]: Maybe call this variable `dependencies`

File: core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java [hunk 9]
```diff
@@ -2124,27 +2143,45 @@ public ResolvedFunction getCoercion(Session session, OperatorType operatorType,
...
+                functionBinding.getFunctionBinding().getBoundSignature());
+        FunctionMetadata functionMetadata = getFunctionMetadata(
+                session,
+                functionBinding.getCatalogHandle(),
+                functionBinding.getFunctionBinding().getFunctionId(),
+                functionBinding.getFunctionBinding().getBoundSignature());
+        return resolve(session, functionBinding.getCatalogHandle(), functionBinding.getFunctionBinding(), functionMetadata, declaration);
+    }
+
+    private FunctionDependencyDeclaration getFunctionDependencyDeclaration(Session session, CatalogHandle catalogHandle, FunctionId functionId, BoundSignature boundSignature)
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java:2173]: Maybe name this `getDependencies`

File: core/trino-main/src/main/java/io/trino/testing/LocalQueryRunner.java [hunk 10]
```diff
@@ -419,6 +411,16 @@ private LocalQueryRunner(
 
         accessControl.setConnectorAccessControlProvider(createAccessControlProvider(connectorManager));
 
+        globalFunctionCatalog.addFunctions(new InternalFunctionBundle(
+                new JsonExistsFunction(functionManager, metadata, typeManager),
+                new JsonValueFunction(functionManager, metadata, typeManager),
+                new JsonQueryFunction(functionManager, metadata, typeManager)));
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/testing/LocalQueryRunner.java:417]: Why is this here?

File: core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java [hunk 11]
```diff
@@ -53,6 +54,7 @@
 
 @ThreadSafe
 public class GlobalFunctionCatalog
+        implements FunctionProvider
 {
     public static final String GLOBAL_SCHEMA = "global";
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java:59]: `builtin`, maybe?

File: core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java [hunk 12]
```diff
@@ -54,7 +54,6 @@
 @ThreadSafe
 public class GlobalFunctionCatalog
 {
-    public static final String GLOBAL_CATALOG = "system";
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/metadata/GlobalFunctionCatalog.java:48]: Can you add some details to the commit explaining how the built in functions are now available for resolution after this change? If I understand it correctly, there's no named catalog for them -- it's just a magic catalog that's always present regardless of what the PATH says, right?

File: core/trino-main/src/main/java/io/trino/testing/LocalQueryRunner.java (hunk 10 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/testing/LocalQueryRunner.java:417]: The json functions use the type system internally so they have to be added after the fact.

**Review by martint — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/function/QualifiedFunctionName.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+import io.trino.spi.Experimental;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+@Experimental(eta = "2022-10-31")
+public class QualifiedFunctionName
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/function/QualifiedFunctionName.java:24]: What's the difference between this class `QualifiedFunctionName` and `io.trino.spi.expression.FunctionName`?
